src/ngx_http_lua_log_ringbuf.c

```diff
@@ -0,0 +1,206 @@
+
+#ifndef DDEBUG
+#define DDEBUG 0
+#endif
+#include "ddebug.h"
+
+
+#include "ngx_http_lua_common.h"
+#include "ngx_http_lua_log_ringbuf.h"
+
+
+typedef struct {
+    unsigned    is_data:1;
+    unsigned    log_level:4;
+    int         len;
+} ngx_http_lua_log_ringbuf_header_t;
+
+
+static void * ngx_http_lua_log_ringbuf_next_header(
+    ngx_http_lua_log_ringbuf_t *rb);
+static void ngx_http_lua_log_ringbuf_append(
+    ngx_http_lua_log_ringbuf_t *rb, int is_data, int log_level, void *buf,
+    int n);
+static size_t ngx_http_lua_log_ringbuf_free_spaces(
+    ngx_http_lua_log_ringbuf_t *rb);
+
+
+void
+ngx_http_lua_log_ringbuf_init(ngx_http_lua_log_ringbuf_t *rb, void *buf,
+    size_t len)
+{
+    rb->data = buf;
+    rb->size = len;
+
+    rb->tail = rb->data;
+    rb->head = rb->data;
+    rb->count = 0;
+    rb->filter_level = NGX_LOG_DEBUG;
+
+    return;
+}
+
+
+void
+ngx_http_lua_log_ringbuf_reset(ngx_http_lua_log_ringbuf_t *rb)
+{
+    rb->tail = rb->data;
+    rb->head = rb->data;
+    rb->count = 0;
+
+    return;
+}
+
+
+/*
+ * get the next data header, it'll skip the useless data space or
+ * placehold data
+ */
+
+static void *
+ngx_http_lua_log_ringbuf_next_header(ngx_http_lua_log_ringbuf_t *rb)
+{
+    ngx_http_lua_log_ringbuf_header_t       *head;
+
+    /* useless data */
+    if (rb->size - (rb->head - rb->data) <
+        sizeof(ngx_http_lua_log_ringbuf_header_t))
+    {
+        rb->head = rb->data;
+        return rb->head;
+    }
+
+    /* placehold data */
+    head = (ngx_http_lua_log_ringbuf_header_t *) rb->head;
+    if (!head->is_data) {
```

Comment:
The `is_data` field is a bit wasteful and confusing. How about introducing a design similar to the "sentinel" thing in nginx's `ngx_rbtree_t` and `ngx_queue_t` data structures?